proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,532 @@
+
+Given that, there are two kinds of potential audience confusion for the token exchange.
+
+1. The ID token issued during the original authorization flow (before token exchange) will have an audience set to the
+   name of the dynamic client. If this client’s name happened to be the same name as the name of a workload cluster,
+   then the client could potentially skip the token exchange and use the original ID token to access the workload
+   cluster via the Concierge (acting as the user who logged in). These initial ID tokens were not meant to grant access
+   to any particular cluster.
+
+2. A user can use the public `pinniped-cli` client to log in and then to perform a token exchange to any audience value.
+   They could even ask for an audience which matches the name of an existing dynamic client to get back an ID token that
+   would appear as if it were issued to that dynamic client. Then they could try to find a way to use that ID token with
+   a webapp which uses that dynamic client to authenticate its users (although that may not be possible depending on the
+   implementation of the webapp).
+
+To address all of these issues we will:
+
+- Require a static, reserved prefix (`client.oauth.pinniped.dev-`) for all dynamic OAuth client IDs.
+- Disallow that prefix (`client.oauth.pinniped.dev-`) for audiences requested via the token exchange API.
+- Disallow the string `pinniped-cli` (the name of the static client) for audiences requested via the token exchange API.
+- Reserve a substring (`.oauth.pinniped.dev`) for audiences requested via the token exchange API for potential future
+  use. Any token exchange requesting an audience with that substring will be rejected. This leaves room to create other
+  categories of audience values in case that is needed in the future, e.g.
+  `static-client.oauth.pinniped.dev-some-client`.
+- All ID tokens issued by the supervisor will contain the `azp` claim and its value will always be set to the client ID
+  of the client that was used for the initial login flow.  This is meant to prevent any information loss during flows
+  such as the token exchange API.
+- No changes are proposed for the `JWTAuthenticator` as it is meant to be interchangeable with the Kubernetes OIDC
+  server flags.  Some environments use old versions of the concierge with newer versions of the supervisor and thus we
+  cannot rely on changes to the concierge being rolled out to enforce security contracts.
+- Implicit behavioral changes to APIs are avoided as they can be difficult to understand and reason about.
+
+The validation associated with dynamic clients will be used to enforce that clients have `metadata.name` set to a value
+that starts with `client.oauth.pinniped.dev-`.  Thus while it will be possible to create a dynamic client CR with a name
+such as `pinniped-cli`, it will not be valid and will not be usable.  For the time being, the hardcoded `pinniped-cli`
```

Comment:
I think the CRD regex prevents creation of such clients.

Reply:
Updated in 7450fb6c8e9937e43fd5f0f95f2caa674413e3eb.